### PR TITLE
[FEATURE] Implement support for event metadata

### PR DESCRIPTION
**Pull request by Akii**
So I've looked into how the metadata thing works with eventstore. And during implementation I fixed some logical problems (fx. that the version is part of the Entry) along with code style inconsistencies.

I marked this as WIP because I also would like to add some more tests. Please let me know what you think about the general approach.

### CODE REVIEW

**Comment by dbellettini** [comment]
Will review this next weekend

File: src/EventStore/EventStore.php [hunk 1]
```diff
@@ -158,7 +157,7 @@ public function writeToStream($stream_name, WritableToStream $events, $expectedV
             )
         ;
 
-        $request->setHeader('ES-ExpectedVersion', intval($expectedVersion));
+        $request->setHeader('ES-ExpectedVersion', intval($expected_version));
```
Comment (dbellettini) [inline src/EventStore/EventStore.php]: yes, we were inconsistent, I prefer camelCase while @nicolopignatelli $expected_version. @nicolopignatelli which rule should we follow? I would use the PSR one...

File: tests/EventStore/Tests/EventStoreTest.php [hunk 2]
```diff
@@ -39,7 +41,7 @@ public function event_is_written_to_stream()
 
     /**
      * @test
-     * @expectedException EventStore\Exception\WrongExpectedVersionException
+     * @expectedException \EventStore\Exception\WrongExpectedVersionException
```
Comment (dbellettini) [inline tests/EventStore/Tests/EventStoreTest.php]: Not sure about the \ at the beginning

**Comment by dbellettini** [comment]
At a first glance looks fine, looks like a BC break but since we're a 0.x it's fine.
I would leave formatting in a separate commit but do it only if your very confident with interactive rebase

@nicolopignatelli can you give your feedback please?

File: src/EventStore/EventStore.php (hunk 1 above)
Comment (nicolopignatelli) [inline src/EventStore/EventStore.php]: Not so important to me. I used underscores to highlight method parameters, but I'm not religious on it.

File: tests/EventStore/Tests/EventStoreTest.php (hunk 2 above)
Comment (Akii) [inline tests/EventStore/Tests/EventStoreTest.php]: PhpStorm complained :)

**Comment by Akii** [comment]
I can separate this PR into two. I must admit I got a little carried away when adding the metadata support.

**Comment by nicolopignatelli** [comment]
Why did you remove all new lines between the opening tag and namespace directive?

File: src/EventStore/StreamFeed/Event.php [hunk 3]
```diff
@@ -9,17 +8,53 @@
...
+    /**
+     * @param array $content
+     */
+    public function __construct(array $content)
+    {
+        $this->version = $content['eventNumber'];
+        $this->type = $content['eventType'];
+        $this->data = $content['data'];
+
+        if ($content['metadata'] !== '') {
```
Comment (nicolopignatelli) [inline src/EventStore/StreamFeed/Event.php]: I would prefer to check for 
```
!empty($content['metadata'])

File: src/EventStore/WritableEvent.php [hunk 4]
```diff
@@ -26,25 +25,33 @@
...
+    private $metadata;
+
+    /**
      * @param  string        $type
      * @param  array         $data
+     * @param  array         $metadata
      * @return WritableEvent
      */
-    public static function newInstance($type, $data)
+    public static function newInstance($type, $data, $metadata = [])
```
Comment (nicolopignatelli) [inline src/EventStore/WritableEvent.php]: add type hint to $data and $metadata

File: src/EventStore/WritableEvent.php [hunk 5]
```diff
@@ -26,25 +25,33 @@
...
-     * @param UUID $uuid
-     * @param $type
-     * @param $data
+     * @param UUID   $uuid
+     * @param string $type
+     * @param array  $data
+     * @param array  $metadata
      */
-    public function __construct(UUID $uuid, $type, $data)
+    public function __construct(UUID $uuid, $type, $data, $metadata = [])
```
Comment (nicolopignatelli) [inline src/EventStore/WritableEvent.php]: add type hint to $data and $metadata

File: tests/EventStore/Tests/EventStoreTest.php [hunk 6]
```diff
@@ -21,7 +21,9 @@ protected function setUp()
         $this->es = new EventStore('http://127.0.0.1:2113');
     }
 
-    /** @test */
```
Comment (nicolopignatelli) [inline tests/EventStore/Tests/EventStoreTest.php]: why transform a 1 line annotation in 3 line?

File: tests/EventStore/Tests/EventStoreTest.php [hunk 7]
```diff
@@ -139,23 +147,53 @@ public function event_stream_feed_returns_entries()
...
         list($entry) = $feed->getEntries();
         $eventUrl    = $entry->getEventUrl();
 
-        $this->assertSame(0, $entry->getVersion());
+        $event       = $this->es->readEvent($eventUrl);
+
+        $this->assertSame(0, $event->getVersion());
+        $this->assertInstanceOf('EventStore\StreamFeed\Event', $event);
+        $this->assertEquals(['foo' => 'bar'], $event->getData());
+        $this->assertSame(NULL, $event->getMetadata());
```
Comment (nicolopignatelli) [inline tests/EventStore/Tests/EventStoreTest.php]: please lowercase null

**Comment by Akii** [comment]
> Why did you remove all new lines between the opening tag and namespace directive?

It was inconsistent across files. I went for what I personally find pretty/more readable.
Same goes for `/** @test */`. There is no need to "save lines". I go for readability, always.

File: src/EventStore/StreamFeed/Event.php [hunk 8]
```diff
@@ -9,17 +8,53 @@
...
-     * @param array $data
-     * @param int   $version
+     * @var array
+     */
+    private $metadata;
+
+    /**
+     * @param array $content
+     */
+    public function __construct(array $content)
```
Comment (nicolopignatelli) [inline src/EventStore/StreamFeed/Event.php]: what do you think about explicitly passing $version, $type and $data instead of an array?

**Comment by nicolopignatelli** [comment]
If you go for readability and don't want to just save lines, then please re-add the newlines after the opening tags.

About the PR in general, seems ok to me.

**Comment by Akii** [comment]
Sorry for the delay, will fix all the commends next week.

**Comment by Akii** [comment]
Still WIP.

**Comment by dbellettini** [comment]
ok, let us know

**Comment by Akii** [comment]
A test for `StreamFeed/EntryTest` isn't really necessary. So from my side, this is now ready for merge.

**Comment by dbellettini** [comment]
ok, tomorrow night I will do a final review and possibly merge

**Comment by Akii** [comment]
Thanks! And don't hesitate to nitpick, I'm not in a hurry so if anything doesn't suit you, just let me know.

**Comment by dbellettini** [comment]
Sorry for the delay, reviewing now

**Pull request merged**